src/goto-symex/symex_goto.cpp

```diff
@@ -547,9 +624,15 @@ void goto_symext::merge_goto(
       "atomic sections differ across branches",
       state.source.pc->source_location);
 
-  if(!goto_state.guard.is_false())
+  // adjust guard, even using guards from unreachable states. This helps to
+  // shrink the state guard if the incoming edge is from a path that was
+  // truncated by config.unwind, config.depth or an assume-false instruction.
+  auto new_guard = std::move(state.guard);
+  new_guard |= goto_state.guard;
```

Comment:
why do you define this new_guard instead of directly modifying state.guard?

Reply:
Added a comment explaining that the `if(!state.reachable)` case overwrites `state.guard`, hence calculating it here but assigning it after the `if(goto_state.reachable)` block.